File: benchmarks/rust/05_matrix_mul.rs
// Matrix Multiplication Benchmark - Rust Implementation
// Multiply two NxN matrices using arrays

fn main() {
    // Benchmark parameter
    let n = 50;
    
    // Initialize matrices
    let mut a = vec![0i64; n * n];
    let mut b = vec![0i64; n * n];
    let mut c = vec![0i64; n * n];
    
    // Fill matrices with values
    for i in 0..n {
        for j in 0..n {
            let idx = i * n + j;
            a[idx] = ((i + j) % 10) as i64;
            b[idx] = ((i * j) % 10) as i64;
        }
    }
    
    // Matrix multiplication: C = A * B
    for i in 0..n {
        for j in 0..n {
            let mut sum = 0;
            for k in 0..n {
                let a_idx = i * n + k;
                let b_idx = k * n + j;
                sum += a[a_idx] * b[b_idx];
            }
            let c_idx = i * n + j;
            c[c_idx] = sum;
        }
    }
    
    // Calculate checksum
    let checksum: i64 = c.iter().sum();
    println!("matrix {}x{} checksum: {}", n, n, checksum);
}

File: benchmarks/python/05_matrix_mul.py
def main():
    n = 50
    a = [0] * (n * n)
    b = [0] * (n * n)
    c = [0] * (n * n)

    i = 0
    while i < n:
        j = 0
        while j < n:
            idx = i * n + j
            a[idx] = (i + j) % 10
            b[idx] = (i * j) % 10
            j += 1
        i += 1

    i = 0
    while i < n:
        j = 0
        while j < n:
            sum_val = 0
            k = 0
            while k < n:
                sum_val += a[i * n + k] * b[k * n + j]
                k += 1
            c[i * n + j] = sum_val
            j += 1
        i += 1

    print(sum(c))


if __name__ == "__main__":
    main()
